### PR TITLE
[move-vm][closures] Types and opcodes for closures

### PR DESCRIPTION
**Pull request by wrwg**
## Description

[PR 1/n for vm closures]

This PR implements the extensions needed in the file format for representing closures:

- The type `SignatureToken` has a new variant `Function(args, result, abilities)` to represent a function type
- The opcodes are extended by new operations `PackClosure`, `PackClosureGeneric`, and `CallClosure`

This supports bit masks for specifying which arguments of a function are captured when creating a closure.

Bytecode verification is extended to support the new types and opcodes. The implementation of consistency, type, and reference safety should be complete. What is missing are:

- Interpreter and value representation
- Value serialization
- Connection of the new types with the various other type representations

## How Has This Been Tested?

Testing postponed until e2e wiring is ready

## Type of Change
- [x] New feature
- [ ] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [x] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Move Compiler
- [ ] Other (specify)

### CODE REVIEW

**Comment by wrwg** [comment]
* **#15667** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/15667?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/15667?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: third_party/move/move-bytecode-verifier/src/reference_safety/mod.rs [hunk 1]
```diff
@@ -559,7 +623,6 @@ fn execute_inner(
     };
     Ok(())
 }
-
```
Comment (georgemitenkov) [inline third_party/move/move-bytecode-verifier/src/reference_safety/mod.rs:562]: Why removing the line?

File: third_party/move/move-bytecode-verifier/src/reference_safety/mod.rs [hunk 2]
```diff
@@ -185,6 +223,18 @@ fn vec_element_type(
     }
 }
 
+fn fun_type(
+    verifier: &mut ReferenceSafetyAnalysis,
+    idx: SignatureIndex,
+) -> PartialVMResult<(Vec<SignatureToken>, Vec<SignatureToken>)> {
+    match verifier.resolver.signature_at(idx).0.first() {
+        Some(SignatureToken::Function(args, result, _)) => Ok((args.clone(), result.clone())),
```
Comment (georgemitenkov) [inline third_party/move/move-bytecode-verifier/src/reference_safety/mod.rs:231]: Do we really need to clone args/returns for `clos_eval`? Seems not necessary

File: third_party/move/move-bytecode-verifier/src/reference_safety/mod.rs [hunk 3]
```diff
@@ -100,6 +101,43 @@ fn call(
...
+    verifier: &mut ReferenceSafetyAnalysis,
+    state: &mut AbstractState,
+    offset: CodeOffset,
+    arg_tys: Vec<SignatureToken>,
+    result_tys: Vec<SignatureToken>,
+    meter: &mut impl Meter,
+) -> PartialVMResult<()> {
+    let arguments = arg_tys
+        .iter()
+        .map(|_| verifier.stack.pop().unwrap())
```
Comment (georgemitenkov) [inline third_party/move/move-bytecode-verifier/src/reference_safety/mod.rs:131]: `safe_unwrap!`?

File: third_party/move/move-bytecode-verifier/src/reference_safety/abstract_state.rs [hunk 4]
```diff
@@ -559,6 +568,16 @@ impl AbstractState {
         Ok(return_values)
     }
 
+    pub fn clos_eval(
```
Comment (georgemitenkov) [inline third_party/move/move-bytecode-verifier/src/reference_safety/abstract_state.rs:571]: Documentation for pubic functions would be nice

File: third_party/move/move-binary-format/src/file_format.rs [hunk 5]
```diff
@@ -1501,21 +1524,119 @@ impl SignatureToken {
...
+    }
+
+    /// Apply a closure mask to a list of elements, returning only those
+    /// where position `i` is set in the mask (if `collect_captured` is true) or not
+    /// set (otherwise).
+    pub fn extract<T: Clone>(&self, tys: &[T], collect_captured: bool) -> Vec<T> {
+        tys.iter()
+            .enumerate()
+            .filter_map(|(pos, x)| {
+                let set = (1 << pos) & self.mask != 0;
```
Comment (georgemitenkov) [inline third_party/move/move-binary-format/src/file_format.rs:1575]: Seems like a bug, if `tys.len() > 64` shift will panic. We should enforce here the number of types is <= 64. I have not seen (or maybe missed) where we enforce this in bytecode verifier.

File: third_party/move/move-bytecode-verifier/src/stack_usage_verifier.rs [hunk 6]
```diff
@@ -228,6 +228,43 @@ impl<'a> StackUsageVerifier<'a> {
...
+                    // ensures we never reach this
+                    (0, 0)
+                }
+            },
+
+            // ClosPack pops the captured arguments and returns 1 value
+            Bytecode::PackClosure(idx, mask) => {
+                let function_handle = self.resolver.function_handle_at(*idx);
+                let arg_count = mask
+                    .extract(
```
Comment (georgemitenkov) [inline third_party/move/move-bytecode-verifier/src/stack_usage_verifier.rs:249]: 1. Why extract, we can probably use `popcount` to count number of 1s in the mask?
2. We need to check that arg_count <= 64 somewhere

File: third_party/move/move-bytecode-verifier/src/stack_usage_verifier.rs [hunk 7]
```diff
@@ -228,6 +228,43 @@ impl<'a> StackUsageVerifier<'a> {
...
+                        true,
+                    )
+                    .len() as u64;
+                (arg_count, 1)
+            },
+            Bytecode::PackClosureGeneric(idx, mask) => {
+                let func_inst = self.resolver.function_instantiation_at(*idx);
+                let function_handle = self.resolver.function_handle_at(func_inst.handle);
+                let arg_count = mask
+                    .extract(
```
Comment (georgemitenkov) [inline third_party/move/move-bytecode-verifier/src/stack_usage_verifier.rs:260]: Same here

File: third_party/move/move-bytecode-verifier/src/signature_v2.rs [hunk 8]
```diff
@@ -881,6 +898,14 @@ impl<'a, const N: usize> SignatureChecker<'a, N> {
                         entry.insert(());
                     }
                 },
+                CallClosure(idx) => {
+                    let sign = self.resolver.signature_at(*idx);
```
Comment (georgemitenkov) [inline third_party/move/move-bytecode-verifier/src/signature_v2.rs:902]: nit: `sign` --> `s`/`signature`/`ty`

File: third_party/move/move-bytecode-verifier/src/signature_v2.rs [hunk 9]
```diff
@@ -881,6 +898,14 @@ impl<'a, const N: usize> SignatureChecker<'a, N> {
...
+                    if sign.len() != 1 || !matches!(&sign.0[0], SignatureToken::Function(..)) {
+                        return map_err(Err(PartialVMError::new(
+                            StatusCode::CLOSURE_EVAL_REQUIRES_FUNCTION,
```
Comment (georgemitenkov) [inline third_party/move/move-bytecode-verifier/src/signature_v2.rs:905]: Add error message? e.g. "Requires function with ABC but got XYZ..."

File: third_party/move/move-bytecode-verifier/src/type_safety.rs [hunk 10]
```diff
@@ -300,6 +301,115 @@ fn call(
...
 
+fn clos_eval(
+    verifier: &mut TypeSafetyChecker,
+    meter: &mut impl Meter,
+    offset: CodeOffset,
+    expected_ty: &SignatureToken,
+) -> PartialVMResult<()> {
+    let SignatureToken::Function(param_tys, ret_tys, abilities) = expected_ty else {
+        // The signature checker has ensured this is a function
+        safe_assert!(false);
```
Comment (georgemitenkov) [inline third_party/move/move-bytecode-verifier/src/type_safety.rs:312]: Why
```
safe_assert!(false);
unreachable!()
```
can't we just return an error? The pattern seems very weird...

File: third_party/move/move-bytecode-verifier/src/type_safety.rs [hunk 11]
```diff
@@ -300,6 +301,115 @@ fn call(
...
+        unreachable!()
+    };
+    // On top of the stack is the closure, pop it.
+    let closure_ty = safe_unwrap!(verifier.stack.pop());
+    // Verify that the closure type matches the expected type
+    if &closure_ty != expected_ty {
+        return Err(verifier
+            .error(StatusCode::CALL_TYPE_MISMATCH_ERROR, offset)
+            .with_message("closure type mismatch".to_owned()));
```
Comment (georgemitenkov) [inline third_party/move/move-bytecode-verifier/src/type_safety.rs:321]: Add more info to the error message? What is the actual and what is the expected types?

File: third_party/move/move-bytecode-verifier/src/type_safety.rs [hunk 12]
```diff
@@ -300,6 +301,115 @@ fn call(
...
+) -> PartialVMResult<()> {
+    let SignatureToken::Function(param_tys, ret_tys, abilities) = expected_ty else {
+        // The signature checker has ensured this is a function
+        safe_assert!(false);
+        unreachable!()
+    };
+    // On top of the stack is the closure, pop it.
+    let closure_ty = safe_unwrap!(verifier.stack.pop());
+    // Verify that the closure type matches the expected type
+    if &closure_ty != expected_ty {
```
Comment (georgemitenkov) [inline third_party/move/move-bytecode-verifier/src/type_safety.rs:318]: How do we implement `Eq` for signature token? Because later you do `!abilities.is_subset(closure_abilities)` but we check for equality here?

File: third_party/move/move-bytecode-verifier/src/type_safety.rs [hunk 13]
```diff
@@ -300,6 +301,115 @@ fn call(
...
+        return Err(verifier
+            .error(StatusCode::CALL_TYPE_MISMATCH_ERROR, offset)
+            .with_message("closure type mismatch".to_owned()));
+    }
+    // Verify that the abilities match
+    let SignatureToken::Function(_, _, closure_abilities) = closure_ty else {
```
Comment (georgemitenkov) [inline third_party/move/move-bytecode-verifier/src/type_safety.rs:324]: Overall, this code is quite confusing, I would prefer it written differently if possible. For example:

```
let closure_ty = safe_unwrap!(verifier.stack.pop());
match (closure_ty, actual_ty) {
  (Function(..), Function(..)) => {
      // do all checks here
  },
  _ => return Err(...)
}
```

File: third_party/move/move-bytecode-verifier/src/type_safety.rs [hunk 14]
```diff
@@ -300,6 +301,115 @@ fn call(
...
+    if !abilities.is_subset(closure_abilities) {
+        return Err(verifier
+            .error(StatusCode::CALL_TYPE_MISMATCH_ERROR, offset)
+            .with_message("closure ability mismatch".to_owned()));
+    }
+    // Pop and verify arguments
+    for param_ty in param_tys.iter().rev() {
+        let arg_ty = safe_unwrap!(verifier.stack.pop());
+        if &arg_ty != param_ty {
+            return Err(verifier.error(StatusCode::CALL_TYPE_MISMATCH_ERROR, offset));
```
Comment (georgemitenkov) [inline third_party/move/move-bytecode-verifier/src/type_safety.rs:338]: Error message with useful details on why there is a mismatch?

File: third_party/move/move-bytecode-verifier/src/type_safety.rs [hunk 15]
```diff
@@ -300,6 +301,115 @@ fn call(
...
+    // Verify that the abilities match
+    let SignatureToken::Function(_, _, closure_abilities) = closure_ty else {
+        // Ensured above, but never panic
+        safe_assert!(false);
+        unreachable!()
+    };
+    if !abilities.is_subset(closure_abilities) {
+        return Err(verifier
+            .error(StatusCode::CALL_TYPE_MISMATCH_ERROR, offset)
+            .with_message("closure ability mismatch".to_owned()));
```
Comment (georgemitenkov) [inline third_party/move/move-bytecode-verifier/src/type_safety.rs:332]: Error message should contain actual and expected abilities, otherwise it is pretty much useless.

File: third_party/move/move-bytecode-verifier/src/type_safety.rs [hunk 16]
```diff
@@ -300,6 +301,115 @@ fn call(
...
+    verifier: &mut TypeSafetyChecker,
+    meter: &mut impl Meter,
+    offset: CodeOffset,
+    func_handle_idx: FunctionHandleIndex,
+    type_actuals: &Signature,
+    mask: ClosureMask,
+) -> PartialVMResult<()> {
+    let func_handle = verifier.resolver.function_handle_at(func_handle_idx);
+    // Check the captured arguments on the stack
+    let param_sign = verifier.resolver.signature_at(func_handle.parameters);
```
Comment (georgemitenkov) [inline third_party/move/move-bytecode-verifier/src/type_safety.rs:357]: please `param_signature` or `param_ty`

File: third_party/move/move-bytecode-verifier/src/type_safety.rs [hunk 17]
```diff
@@ -300,6 +301,115 @@ fn call(
...
+            return Err(verifier.error(StatusCode::CALL_TYPE_MISMATCH_ERROR, offset));
+        }
+    }
+    for ret_ty in ret_tys {
+        verifier.push(meter, ret_ty.clone())?
+    }
+    Ok(())
+}
+
+fn clos_pack(
```
Comment (georgemitenkov) [inline third_party/move/move-bytecode-verifier/src/type_safety.rs:347]: Can we unit test this?

File: third_party/move/move-bytecode-verifier/src/type_safety.rs [hunk 18]
```diff
@@ -300,6 +301,115 @@ fn call(
...
+        let arg = safe_unwrap!(verifier.stack.pop());
+        if (type_actuals.is_empty() && &arg != ty)
+            || (!type_actuals.is_empty() && arg != instantiate(ty, type_actuals))
+        {
+            return Err(verifier
+                .error(StatusCode::PACK_TYPE_MISMATCH_ERROR, offset)
+                .with_message("captured argument type mismatch".to_owned()));
+        }
+        // A captured argument must not be a reference
+        if ty.is_reference() {
```
Comment (georgemitenkov) [inline third_party/move/move-bytecode-verifier/src/type_safety.rs:371]: I am not very familar with the codebase, but I think `!ty.is_value()` might have stronger guarantees?

Comment (wrwg) [reply third_party/move/move-bytecode-verifier/src/type_safety.rs:374]: Doesn't exist on SignatureToken

File: third_party/move/move-bytecode-verifier/src/type_safety.rs (hunk 10 above)
Comment (georgemitenkov) [reply third_party/move/move-bytecode-verifier/src/type_safety.rs:312]: Didn't see the definition of `safe_assert` is to panic with debug assertions, I thought it is always an error. I think the name could have been better, e.g., `debug_panic_or_error()`, but whatever

File: third_party/move/move-bytecode-verifier/src/instruction_consistency.rs [hunk 19]
```diff
@@ -227,4 +237,19 @@ impl<'a> InstructionConsistency<'a> {
...
+    fn check_closure_mask(
+        &self,
+        offset: usize,
+        func_handle_index: FunctionHandleIndex,
+        mask: ClosureMask,
+    ) -> PartialVMResult<()> {
+        let function_handle = self.resolver.function_handle_at(func_handle_index);
+        let signature = self.resolver.signature_at(function_handle.parameters);
+        if mask.max_captured() >= signature.len() {
+            return Err(PartialVMError::new(StatusCode::INVALID_CLOSURE_MASK)
```
[the remark this answers is not in the document]
Comment (georgemitenkov) [reply third_party/move/move-bytecode-verifier/src/instruction_consistency.rs:249]: Maybe add a comment about that? Seems like there are many indirect assumptions which is difficult to see/know when reviewing/coding in unfamiliar codebase

File: third_party/move/move-bytecode-verifier/invalid-mutations/src/bounds/code_unit.rs [hunk 20]
```diff
@@ -498,6 +498,9 @@ impl<'a> ApplyCodeUnitBoundsContext<'a> {
                         // TODO(#13806): implement
                         panic!("Enum types bytecode NYI: {:?}", code[bytecode_idx])
```
[the remark this answers is not in the document]
Comment (georgemitenkov) [reply third_party/move/move-bytecode-verifier/invalid-mutations/src/bounds/code_unit.rs:499]: Yes, we should probably come back to that otherwise it is difficult once the feature is out (I myself did not address all todos with aggregators and loader... yet😄)

File: third_party/move/tools/move-disassembler/src/disassembler.rs [hunk 21]
```diff
@@ -570,6 +570,9 @@ impl<'a> Disassembler<'a> {
         type_param_context: &[SourceName],
     ) -> Result<String> {
         Ok(match sig_tok {
+            // TODO: function types
```
Comment (georgemitenkov) [inline third_party/move/tools/move-disassembler/src/disassembler.rs:573]: Add github issue?

File: third_party/move/move-binary-format/src/file_format.rs [hunk 22]
```diff
@@ -1501,21 +1524,119 @@ impl SignatureToken {
...
+    pub mask: u64,
+}
+
+impl fmt::Display for ClosureMask {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{:b}", self.mask)
+    }
+}
+
+impl ClosureMask {
```
[the remark this answers is not in the document]
Comment (georgemitenkov) [reply third_party/move/move-binary-format/src/file_format.rs:1563]: Well, it was not simple here (compared to your next PR) and had implicit assumptions which were not documented or tested 😄 Since you rewrite this part in any case, it is ok to land as is in this PR

**Comment by georgemitenkov** [comment]
@wrwg I think this PR got V8 binary format change leaked into it

**Comment by wrwg** [comment]
> @wrwg I think this PR got V8 binary format change leaked into it

Can't see any unexpected. This PR is based on the (now landed) v8 PR. Perhaps there was a merge conflict which is now resolved after rebase?

**Review by georgemitenkov — APPROVED** [review]
Let's land this!

P.S. Let's remember to add tests for corner cases once things are connected e2e, e.g. creating a closure for function with > 64 args, etc.

File: third_party/move/move-bytecode-verifier/src/signature_v2.rs [hunk 23]
```diff
@@ -173,6 +173,18 @@ fn check_ty<const N: usize>(
                 param_constraints,
             )?;
         },
+        Function(args, result, abilities) => {
+            assert_abilities(*abilities, required_abilities)?;
+            for ty in args.iter().chain(result.iter()) {
+                check_ty(
```
Comment (runtian-zhou) [inline third_party/move/move-bytecode-verifier/src/signature_v2.rs:179]: Do we actually need to check the arugments abilities here? The arguments ability should be irrelavant to the closure's ability?

File: third_party/move/move-binary-format/src/file_format.rs [hunk 24]
```diff
@@ -2935,6 +3042,84 @@ pub enum Bytecode {
...
+        ```
+          CallClosure(PackClosure(f, mask, c1..cn), a1..am) ==
+             f(mask.compose(c1..cn, a1..am))
+        ```
+    "#]
+    #[static_operands = "[]"]
+    #[semantics = ""]
+    #[runtime_check_epilogue = ""]
+    #[gas_type_creation_tier_0 = "closure_ty"]
+    CallClosure(SignatureIndex),
```
Comment (runtian-zhou) [inline third_party/move/move-binary-format/src/file_format.rs:3121]: The signature index here refers to the types of values on the stack?

File: third_party/move/move-bytecode-verifier/src/signature_v2.rs [hunk 25]
```diff
@@ -881,6 +898,15 @@ impl<'a, const N: usize> SignatureChecker<'a, N> {
                         entry.insert(());
                     }
                 },
+                CallClosure(idx) => {
+                    let sgn = self.resolver.signature_at(*idx);
```
Comment (runtian-zhou) [inline third_party/move/move-bytecode-verifier/src/signature_v2.rs:902]: This should be checked by the type checker as well right?

File: third_party/move/move-bytecode-verifier/src/type_safety.rs (hunk 12 above)
Comment (runtian-zhou) [reply third_party/move/move-bytecode-verifier/src/type_safety.rs:318]: Yea I would be a bit concerned about the usability of code around co-variance here. The code here will be a lot more painful to modify after we ship the feature so really want us to have a clear story here. Will the lack of covariance here actually hurt the usability?

File: third_party/move/move-bytecode-verifier/src/type_safety.rs [hunk 26]
```diff
@@ -300,6 +301,121 @@ fn call(
...
+    let func_handle = verifier.resolver.function_handle_at(func_handle_idx);
+    // Check the captured arguments on the stack
+    let param_sgn = verifier.resolver.signature_at(func_handle.parameters);
+    let captured_param_tys = mask.extract(&param_sgn.0, true);
+    let mut abilities = AbilitySet::ALL;
+    for ty in captured_param_tys.into_iter().rev() {
+        abilities = abilities.intersect(verifier.abilities(ty)?);
+        let arg = safe_unwrap!(verifier.stack.pop());
+        if (type_actuals.is_empty() && &arg != ty)
+            || (!type_actuals.is_empty() && arg != instantiate(ty, type_actuals))
```
Comment (runtian-zhou) [inline third_party/move/move-bytecode-verifier/src/type_safety.rs:367]: Why do you need the `||` here? Just for performance?

**Review by runtian-zhou — COMMENTED** [review]
Just did a scan over the bytecode verifier!

File: third_party/move/move-binary-format/src/file_format.rs [hunk 27]
```diff
@@ -2935,6 +3042,84 @@ pub enum Bytecode {
...
+        Notice that an implementation can derive the types of the captured arguments
+        at runtime from a closure value as long as the closure value stores the function
+        handle (or a derived form of it) and the mask, where the handle allows to lookup the
+        function's type at runtime. Then the same procedure as outlined above can be used.
+    "#]
+    #[static_operands = "[fun, mask]"]
+    #[semantics = ""]
+    #[runtime_check_epilogue = ""]
+    #[gas_type_creation_tier_0 = "closure_ty"]
+    PackClosure(FunctionHandleIndex, ClosureMask),
```
[the remark this answers is not in the document]
Comment (wrwg) [reply third_party/move/move-binary-format/src/file_format.rs:3081]: Interesting idea, but the problem is since this set would be a 'claim', the bytecode and runtiime verifier still need to construct the actual provided abilities from the parameters to verify its true. Therefore we just better leave it out, I think.

File: third_party/move/move-bytecode-verifier/src/signature_v2.rs (hunk 23 above)
Comment (wrwg) [reply third_party/move/move-bytecode-verifier/src/signature_v2.rs:179]: Good finding! Yes, you are right, this check is unnecessary. Removed.

File: third_party/move/move-bytecode-verifier/src/type_safety.rs (hunk 26 above)
Comment (wrwg) [reply third_party/move/move-bytecode-verifier/src/type_safety.rs:367]: Yeah, don't want to run the instantiate until needed.

**Review by wrwg — COMMENTED** [review]
Thanks for the reviews!

File: third_party/move/move-core/types/src/vm_status.rs [hunk 28]
```diff
@@ -868,11 +872,14 @@ pub enum StatusCode {
     // Struct variant not matching. This error appears on an attempt to unpack or borrow a
     // field from a value which is not of the expected variant.
     STRUCT_VARIANT_MISMATCH = 4038,
+    // An unimplemented feature in the VM.
+    UNIMPLEMENTED_FEATURE = 4039,
```
[the remark this answers is not in the document]
Comment (wrwg) [reply third_party/move/move-core/types/src/vm_status.rs]: I see now where the confusion comes from, its 'feature' in 'feature flags' vs 'feature' generically. Renamed the error into UNIMPLEMENTED_FUNCTIONALITY

**Review by runtian-zhou — COMMENTED** [review]
The logics looks good to me overall as a first iteration. Couple of thoughts:
1. I'm wondering how much would the lack of variance might hurt in our later production. This might be a complicated issue as we need to determine the abilities for the stored closures as well. So I cannot just eyeball how bad the situation might be and if the issue could be resolved by compiler, or do we need an extra opcode to handle the variance, similar to `FreezeRef`.
2. As we are landing this to main, we need to make sure the feature is gated properly. This PR itself seems fine because we are only adding new binary instructions and these are gated by the VM version that we are well aware of. However, the next PR that introduces changes to TypeTag woluld be really tricky, as TypeTag would be used externally everywhere. TypeTag is both used in our txn format and by the resource viewer, so rolling out the next PR would be a lot harder than this one IMO.
3. We will need closer eyes on the bytecode verifiers. Should we start the auditing process once we merge this PR rightaway?

File: third_party/move/move-bytecode-verifier/src/type_safety.rs (hunk 12 above)
Comment (wrwg) [reply third_party/move/move-bytecode-verifier/src/type_safety.rs:318]: Folks, these are all standard patterns. Rust doesn't have co or contra variance, and multiple other languages too. Still all have substitutability/assignability. Please read the AIP also for discussion.

Rust e.g. considers `R<&T>` and `R<&mut T>` completely different types -- even though it allows to assign `&mut T` to `&T`.

Also not adding co-variance now will not block us from adding it later if we really wanted.

Comment (runtian-zhou) [reply third_party/move/move-bytecode-verifier/src/type_safety.rs:318]: I was more talking about the code here. If we already checked `closure_ty == expected_ty`, by using the default Eq implementation we are requiring the ability of the two closure types to be the same as well, which is not what you would want for assignability?

Comment (wrwg) [reply third_party/move/move-bytecode-verifier/src/type_safety.rs]: Indeed! I have forgotten to update this file, its still in the state from October.

Now introduced `token.is_assignable_from(other)` and applied at the appropriate locations here. Also documented where should not be applied and equality used instead.

Also removed the unnecessary ability check in `call_closure`.

**Review by runtian-zhou — COMMENTED** [review]
Was wondering if the current interaction between vector operation and the assignability. One interesting outcome is we can now push a closure with drop ability into a vector of same type without drop ability. However when we popped the item, it will be treated as a non droppable value from the verifier standpoint. Was wondering if we should make vector operations to allow no variance/strict assignablity at all to make things easier.

**Comment by wrwg** [comment]
This is a common property in languages so I'm not too concerned about it. Consider Rust:

```
let v: Vec<&usize> = vec![];
let x: &mut usize = ..;
v.push_back(x);
```

This is allowed, even though when you pop the value, it will be only a `&usize`.

The general pattern, which you find in multiple languages with generics, is that you have shallow variance (the 'is_asssignable' property) but not general variance. I don't think it causes any unexpected problems.

**Review by runtian-zhou — APPROVED** [review]
LGTM! Let's get ppl unblocked!
